DFAST-44 : Create developer starters guide (#5)
Deltares/D-FAST_Morphological_Impact: docs/chapters/developer_guide.tex

> \chapter{Developer Guide}

Reviewer: Overall comment:
Maybe good if we had a look together in what order we would expect the developer guide to go.
e.g.
# setting up enviroment.
# running a test
# running coverage
# debugging a test
# information regarding dependencies (poetry & toml)
# adding/updating dependencies
# Nuitka build?

Author: see DFAST-132

\section{Introduction}

This chapter gives an overview of how \dfastmi can be setup for development.

> 

Author: Yes, processed in document

> The development setup and configuration is done for usage of the visual studio code IDE. More information about visual studio code can be found on the internet.

Reviewer: maybe add  the link instead of saying "on the internet"?
https://code.visualstudio.com/

Author: agree, processed

> The development of python code project within Deltares is mostly done in with this tool. Visual Studio Code is free for private or commercial use. See the \href{https://code.visualstudio.com/license}{product license}) for details.

Reviewer: ```suggestion
The development of python code project within Deltares is mostly done with this tool. Visual Studio Code is free for private or commercial use. See the \href{https://code.visualstudio.com/license}{product license}) for details.
```

\begin{Remark}
	\item This configuration can not be used to build a standalone executable with nuitka tooling. It is for development purposes only.
\end{Remark}

\section{Configuration}

> The tool can be configured with a batch script which will configure using installed tools or ask the user to download and install the tools which can be used for them. The batch file is located in the sub folder \textbf{BuildScripts} and is called \textbf{DevelopDfastmi.bat}.

Reviewer: ```suggestion
The DFAST-MI developer environment can be configured with a batch script. This script will configure and install recommended tools. User input is required to download and install these tools. The batch file is located in the sub folder \textbf{BuildScripts} and is called \textbf{DevelopDfastmi.bat}.
```

The batch file will check for installed tooling for anaconda/miniconda. If not installed it will ask the user to install the anaconda/miniconda tooling for you. 

> \begin{Remark}

Reviewer: ```suggestion

\begin{Remark}
```

Reviewer: Remark is now directly behind the sentence instead of a new line.

![fA0pkiK0n1](https://github.com/Deltares/D-FAST_Morphological_Impact/assets/102953403/aeb0973a-484b-4e60-acc0-5a84aa9f5920)

	\item \textbf{PLEASE NOTE:} you need to close the command line and re-open / restart the batch script as it needs to initialize the conda tooling for the command line and powershell (used in VSCode).
\end{Remark}

> The batch file will check for installed tooling for VSCode. If not installed it will ask the user to install the VSCode tooling for you. 

Reviewer: Maybe add that vscode extensions will be installed?

Author: yup, done.

\begin{Remark}
	\item \textbf{PLEASE NOTE:} you need to close the command line and re-open / restart the batch script as it needs to initialize the VS Code tooling for the command line.
\end{Remark}

\section{Configuration files}

> As stated earlier we use conda to create a clean python environment to develop the \dfastmi application in. We also use poetry to setup the needed packages for the development environment. Poetry uses a toml\footnote{TOML : Tom's Obvious Minimal Language. A config file format file} file for the configuration (\textbf{poetry.toml}) and a lock file (\textbf{poetry.lock}) which i have no clue why it is there. Also we use a \textbf{setup.py} file used by the setup tools package to setup the application for development. And last but not least a python tools / interpreter configuration \textbf{pyproject.toml}. This file configures the way python packages should behave.

Reviewer: Would you like to leave the following sentence in the developers guide?

```
A config file format file} file for the configuration (\textbf{poetry.toml}) and a lock file (\textbf{poetry.lock}) which i have no clue why it is there.
```

Reviewer: If you want to keep this sentence, please use a capital "I"

Reviewer: For more information on the poetry.lock file, see the following page:
https://python-poetry.org/docs/basic-usage/#installing-without-poetrylock

Author: I think i worded it now better.

> \subsection{Conda environment}

Reviewer: I am not sure what you want to convey in the "Conda environment" chapter

Author: I added explanation about the isolated environment to get reproducible development environments

> We setup a conda environment on the client pc so we have the python interpreter we expect to be used by developers.

Reviewer: what do you exactly refer to with "client pc"?

Author: i have remove that part

\begin{table}[]

> 	\caption{Default packages for conda environment with python interpreter v3.9.13.}

Reviewer: These packages are default installed by the conda installer?

Author: Yes, I think i worded it better now.

	\begin{tabular}{|l|l|}
		\hline
		\textbf{Package}    & \textbf{Remark} \\ \hline
		ca-certificates     &                 \\ \hline
		openssl             &                 \\ \hline
		pip                 &                 \\ \hline
		python              &                 \\ \hline
		setuptools          &                 \\ \hline
		sqlite              &                 \\ \hline
		tzdata              &                 \\ \hline
		vc                  &                 \\ \hline
		vs2015\_runtime     &                 \\ \hline
		wheel               &                 \\ \hline
	\end{tabular}
\end{table}

> \subsubsection{Conda / Python packages needed for \dfastmi}

Reviewer: Where did you get these packages from?

Aren't the needed packages stated in the PyProject.toml and poetry.lock file?

Reviewer: Is there a specific reason to add these packages in a list in the user guide?

	platformdirs (4.1.0)          &                 \\ \hline
	pyparsing (3.1.1)             &                 \\ \hline
	pyproj (3.6.1)                &                 \\ \hline
	pyqt5-qt5 (5.15.2)            &                 \\ \hline
	pyqt5-sip (12.13.0)           &                 \\ \hline
	pytest (7.4.4)                &                 \\ \hline
	shapely (2.0.2)               &                 \\ \hline
	tbb (2021.7.1)                &                 \\ \hline
	typing-extensions (4.9.0)     &                 \\ \hline
	zstandard (0.22.0)            &                 \\ \hline
	black (22.12.0)               &                 \\ \hline
	geopandas (0.14.2)            &                 \\ \hline
	matplotlib (3.8.2)            &                 \\ \hline
	mkl (2021.4.0)                &                 \\ \hline
	nuitka (1.9.7)                &                 \\ \hline
	pyqt5 (5.15.10)               &                 \\ \hline
	netcdf4 (1.6.5)               &                 \\ \hline
	pytest-cov (4.1.0)            &                 \\ \hline
	teamcity-messages (1.32)      &                 \\ \hline
\end{longtable}

\section{Tooling}

> The tool is using:

Reviewer: ```suggestion
DFAST Morphological Impact is using:
```

\begin{enumerate}
\item \keyw{anaconda} / \keyw{miniconda} /\keyw{conda} to setup its python environment. 
\item \keyw{poetry} to install the python packages for the environment
\item \keyw{VSCode} to edit python files, run the created (unit)tests, see/visualize the test coverage
\end{enumerate}

\subsection{Utility / install scripts}

> We use several other batch scripts.

Reviewer: Might this chapter be more part of the Configuration chapter?

Author: for now i want to keep it here, after merge we will get probably same or similar feedback from Bert. Lets get a chapter with sections in the user manual and later decide to move / replace where needed. Good idea, but lets progress first.

\begin{itemize}
	\item \keyw{CondaInstall.bat} used to install miniconda on the client pc, you need to restart the main script (DevelopDfastmi.bat) in a \textbf{new} command line prompt because the command line prompt environment is
	 updated after install.
	\item \keyw{VSCodeInstall.bat} used to install visual studio code on the client pc, you need to restart the main script (DevelopDfastmi.bat) in a \textbf{new} command line prompt because the command line prompt environment is updated after install.
\end{itemize}

\subsection{VSCode Extensions}
In VSCode we use extensions to find python unit test, visualize coverage and debug our code. To do this we install the following extensions.

> \begin{longtable}{|l|l|}

Reviewer: Maybe good to add a short remark on what the extension does and why it is needed?

Author: yes, did that.

	\hline
	\textbf{VS Code extension} & \textbf{Remark} \\ \hline
	\endhead
	%
	Cameron.vscode-pytest                 		&                 \\ \hline
	donjayamanne.python-environment-manager     &                 \\ \hline
	hbenl.vscode-test-explorer                 	&                 \\ \hline
	littlefoxteam.vscode-python-test-adapter    &                 \\ \hline
	ms-python.pylint                 			&                 \\ \hline
	ms-python.python                 			&                 \\ \hline
	ms-python.vscode-pylance                 	&                 \\ \hline
	ms-vscode.test-adapter-converter            &                 \\ \hline
	ryanluker.vscode-coverage-gutters           &                 \\ \hline
\end{longtable}

\subsection{VSCode userfiles}
The user files can be found in subfolder \textbf{.vscode}
\begin{enumerate}
	\item \keyw{extensions.json} will reference the expected and recommended visual studio code extensions. It will not install them but inform the user what is missing to start developing.
	\item \keyw{launch.json} will configure VS Code what to do when an user will debug unit tests or start debugging the application.

> 	\item \keyw{settings.json} configured settings for 

Reviewer: this sentence is not finished.

Author: finished

> 	\item \keyw{settings.json} configured settings for 

Reviewer: ```suggestion
	\item \keyw{settings.json} configured settings for visual studio code.
```

\end{enumerate}

> \subsection{Python/TOML}

Reviewer: ```suggestion
\subsection{TOML}
```

> This file is used by the python interpreter to configure the application. The TOML file itself has nothing to do with python, poetry or VS Code but it is a configuration file which is easy to read.

Reviewer: This text does not seems correct and it seems in contradiction with what you stated earlier in chapter Configuration files.
The start of this text refers to a specific file, but which file is not fully clear, only that it is a toml file.

> This file is used by the python interpreter to configure the application. The TOML file itself has nothing to do with python, poetry or VS Code but it is a configuration file which is easy to read.

Reviewer: ```suggestion
This filetype is used by the python interpreter to configure the application. The TOML file itself has nothing to do with python, poetry or VS Code but it is a configuration file which is easy to read.
```

> Added configuration settings to the \textbf{pyproject.toml} file

Reviewer: Might be good to also refer to the documentation related to these two?
(https://coverage.readthedocs.io/en/latest/config.html#toml-syntax)

Author: good remark, added to file
